sklearn/ensemble/tests/test_gradient_boosting.py

```diff
@@ -1324,3 +1324,15 @@ def test_gradient_boosting_validation_fraction():
     gbr3.fit(X_train, y_train)
     assert gbr.n_estimators_ < gbr3.n_estimators_
     assert gbc.n_estimators_ < gbc3.n_estimators_
+
+
+def test_early_stopping_stratified():
+    # Make sure data splitting for early stopping is stratified
```

Comment:
This test is rather weak, since it relies on the fact that this error message is only raised by `StratifiedShuffleSplit`.
I suggest:
```py
X = [[1, 2], [2, 3], [3, 4], [4, 5]]
y = [0, 0, 1, 1]
gbc = GradientBoostingClassifier(validation_fraction=0.5, n_iter_no_change=5)
# some non-stratified random splits would select only one class in y,
# leading to an error
for _ in range(100):
    gbc.fit(X, y)
```

Reply:
I'm not sure about this, this test still has a (very low) probability to not fail just by chance even if the split isn't stratified.

Also it wouldn't work for MLP

Reply:
But I agree mine isn't great either... no strong opinion

Reply:
1 / 3**100 = 1e-48, I think the probability is low enough.
Why wouldn't it work for MLP ?

Reply:
Because MLP wouldn't raise an error whether splits are stratified or not. The error is only raised by GBDTs because they use an `init` estimator (and that's only once #12983 is merged)

Reply:
`GradientBoostingClassifier` does raise an error when fitting on only one class.
`MLPClassifier` does not indeed, which could probably be considered as a bug.

Reply:
It's not just about raising an error when training on only one class.

It's about training the GBDT on C classes, and training the init estimator on C - X classes, where X is the number of classes that aren't present after the (non-stratified) split

Reply:
Hmm wait I thought `train_test_split` would raise an error if it coudn't populate y_train with at least one sample belonging to each class but that's not the case

So this PR doesn't fix my original issue...

(I still think splits should be stratified though)

Reply:
What would you think about a test that makes sure the GBDT and MLP can predict both classes on a very imbalances dataset?

I'd design the test to fail on master but pass on this branch, so that'd be some kind of non-regression test.

Reply:
I agree splits should be stratified, both for binary and multiclass problems, to preserve imbalanced class distributions. On the extreme, it could even lead to training only on a subset of classes, which can be problematic.

My suggested test is indeed a non-regression test, checking that both estimators use stratified splits on a over-simplistic 2-class toy problem. It relies on the fact that using a (non-stratified) split which would happen to take only 1 class in the training set would lead to an error. The error is already raised in `GradientBoostingClassifier` but not in `MLPClassifier`, which I consider as a bug.

Reply:
Well until this "bug" is fixed this test is not a non-regression test.

Also, even with such a small probability, it still feels weird to me and I'm not sure that's good practice. But I'm happy to hear what other have to say